### ISSUE TITLE
RecvStream::is_end_stream() shouldn't return true after encountering error

### ISSUE THREAD

**Issue by eaufavor** [hyperium/h2#806]
From the [doc](https://docs.rs/h2/latest/h2/struct.RecvStream.html#method.is_end_stream)

> Returns true if the receive half has reached the end of stream.
> A return value of true means that calls to poll and poll_trailers will both return None.

However in reality, it would return true also after the connection / stream encounters errors because 
https://github.com/hyperium/h2/blob/77be6648e5c6da22f51e5049347496335a032288/src/proto/streams/recv.rs#L560-L562
and `is_recv_closed` is true regardless of the actual cause of the closure
https://github.com/hyperium/h2/blob/77be6648e5c6da22f51e5049347496335a032288/src/proto/streams/state.rs#L416-L420

The problem of `RecvStream::is_end_stream()` returning true after encountering error is that it hints the user that the http request finishes successfully. So complete responses might be interpreted as complete ones.

My suggestion is that this function should only return true after seeing END_STREAM flag.

**Comment by seanmonstar** [comment]
> The problem of RecvStream::is_end_stream() returning true after encountering error is that it hints the user that the http request finishes successfully. So incomplete responses might be interpreted as complete ones.

I think you're right. Want to submit a pull request fixing this?

**Comment by eaufavor** [comment]
Yes will do. Thanks for the confirmation.

**Comment by shikhar** [comment]
We started running into this bug in turmoil-based tests – connection errors like `BrokenPipe` being suppressed and returned as an empty streaming response by tonic. This only started happening after a tonic upgrade to 0.13, because tonic started relying on the documented `is_end_stream()` API contract https://github.com/hyperium/tonic/commit/60b131d2cd066dc0ee386175bf7c7124c23c72f2#diff-382c4b4077926c72746fb47f766ed83187264a4b101a42bff90fbfde757d0314R38 (cc @tottoto FYI)

I verified that the proposed PR fixes the problem. Thank you @eaufavor!

@seanmonstar it'd be great if you are able to merge and release this bugfix soon

**Comment by shikhar** [comment]
hi @seanmonstar ! We currently have a git dependency on h2 to pull this in, if it's not too much trouble a bugfix release would be helpful as it likely affects other tonic users too.

**Comment by seanmonstar** [comment]
Yes you're right, sorry. I'll get a release out on Monday (Friday releases ruin weekends).

**Comment by LucioFranco** [comment]
@shikhar (or others) once the h2 release is out feel free to cut a PR to bump tonics requirement for h2 and I will cut a minor release with that.

**Comment by seanmonstar** [comment]
Published v0.4.10.

(I don't think tonic needs to anything, it's a patch release in a dependency, anyone can just `cargo update`.)
